content/blog/2020-10-28-type-based-alias-analysis.md

```diff
@@ -0,0 +1,138 @@
+performance benefits, and don’t lend themselves to comparing two different 
+alias analyses – in both cases, it’s hard to tell whether the disambiguated 
+pointers will be relevant for the intended use case. Dynamic properties, on the 
+other hand, suffer from overspecificity – it’s difficult to evaluate general 
+efficacy of an alias analysis from performance improvements on a few specific 
+optimizations and benchmarks. Worst of all, neither evaluation can give an idea 
+of how much better an analysis could be.
+
+To alleviate these problems, the paper introduces the concept of limit 
+evaluation to figure out whether there were missed optimization opportunities 
+due to undetected aliases. The authors instrumented their benchmarks to record 
+the address and value of every load, allowing them to realize exactly how many 
+aliases there actually are in an execution. This extra step indicated that the 
+TBAA-assisted RLE removed between 37% and 87% of redundant loads, and for 6 out 
+of 8 of the benchmarks only 5% of the remaining loads were redundancies that 
+could be eliminated. Pushing further, the authors even noted that, in the other 
+two benchmarks, the majority of the remaining loads were due to limitations in 
+the RLE implementation, and not a single missed opportunity was due to TBAA 
+failing to disambiguate references. 2.5% of the remaining loads were due to 
+unknown causes, making that an upper bound for the possible improvement.
+
+Ultimately, the evaluations show that though it is easy to come up with 
+examples in which TBAA fails to properly differentiate references, in practice 
+it has very little room for improvement with respect to redundant load 
+elimination. By evaluating on four metrics, the authors were able to draw more 
+nuanced conclusions; for example, a runtime-only evaluation might conclude that 
+_TypeDecl_ is sufficient, but _FieldTypeDecl_ actually yields significantly 
+more opportunities for RLE. The authors also do not state their results in a 
+vacuum; they clearly state that the results are only with respect to one 
+optimization (RLE) and the set of benchmarks they used. However, benchmark 
+diversity and, in this case, optimization diversity is still lacking; the 
+evaluation metrics reveal deep insight about this case, but are not broadly 
+applied.
+
+The significance of this work is not to be understated; alias analyses over 
+types instead of instructions run much more efficiently due to the vastly 
+reduced search space, and the results in paper indicate that the precision 
+tradeoff could be minimal for some applications. From a modern perspective, 
+it seems that compiler developers have come to the same conclusion; many 
+modern compilers implement and support type-based alias analyses for many of 
+their optimization passes. Among these compilers are GCC and LLVM, showing that 
+the benefits have even extended outside the paper’s original realm of type-safe 
+languages. 

```

Comment:
No need to add anything here, but I think it's interesting that the reason this stuff works in C/C++ is because the standards make it illegal to create two differently-typed references to the same value. Nothing actually stops programs from doing that and lots of software out there does it, of course, but all that software is broken, which is why TBAA is allowed.

Reply:
ah! i was wondering what the breakthrough could have been for this to work outside type safety, but this seems a lot simpler, haha